### ISSUE TITLE
Exact FHMM issues

### ISSUE THREAD

**Issue by nipunbatra** [nilmtk/nilmtk#67]
I implemented exact FHMM. It is on similar lines as suggested by @oliparson in his 9 month report. Unfortunately, it consumes too much memory and hangs the system. On further analysis, if we have K=2, n=15

Memory required for Pi: K^N
Memory required for A: K^N * K^N
Memory required for B (phi): K^N

Total = O(K^2N)

If each variable takes 32 bytes, then atleast:

32*2^30 bytes = 34359738368 byes = 34 GB

Clearly, this won't work! A lot of effort since the previous few months seems useless now!

@oliparson : Did you have similar experiences? 

The model works awesome for ~4 appliances. Didn't test for more currently. I am sure for upto 7-8 appliances it should work great.

Useful references:

1. http://www.cs.ubc.ca/~murphyk/Papers/dbntalk.pdf

**Comment by nipunbatra** [comment]
Some point in the near future, I'd like to implement a 'better' FHMM. Mingjun told me that he used Mosek and implemented FHMM as an optimisation problem. Probably, we can choose something like cvxpy and implement the same.

Also happened to notice this Bayesian HMM implementation: https://github.com/bhmm/bhmm
@oliparson : can this BHMM implementation work similar to your C# Bayesian HMM code?

**Comment by JackKelly** [comment]
Sounds like a great idea.

It might be worth keeping the existing FHMM implementation and adding a new
one (perhaps with a slightly different name) unless we're absolutely
certain that the new implementation will provide exactly the same results.
The reason is that people have used the existing FHMM implementation as a
benchmark so it might be good to allow uses to access this same benchmark
if they want

**Comment by nipunbatra** [comment]
Yes, not planning to remove the existing one. Just a different way of solving (exact vs appx)

**Comment by PMeira** [comment]
I'm adding this issue to the v0.3 milestone to remember to update the docs and maybe add a simple approximate check if the machine running NILMTK even has enough memory to address N appliances.
